internal/rfc7512/rfc7512.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	// Parse the path attributes:
+	for _, attr := range pAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if len(parts) == 2 {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		} else {
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Path Attribute: %s", parts[0])
+		}
+		switch parts[0] {
+		case "token":
+			pk11uri.Token = value
+		case "manufacturer":
+			pk11uri.Manuf = value
+		case "serial":
+			pk11uri.Serial = value
+		case "model":
+			pk11uri.Model = value
+		case "library-manufacturer":
+			pk11uri.LibManuf = value
+		case "library-description":
+			pk11uri.LibDesc = value
+		case "library-version":
+			// TODO 1*DIGIT [ "." 1*DIGIT ]
+			pk11uri.LibVer = value
+		case "object":
+			pk11uri.Object = []byte(value)
+		case "type":
+			// TODO public, private, cert, secret-key, data
+			pk11uri.Type = value
+		case "id":
+			pk11uri.ID = []byte(value)
+		case "slot-manufacturer":
+			pk11uri.SlotManuf = value
+		case "slot-description":
+			pk11uri.SlotDesc = value
+		case "slot-id":
+			// TODO the bit size is not clarified.
+			id, _ := strconv.ParseUint(value, 10, 32)
+			pk11uri.SlotID = uint(id)
+		default:
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Path Attribute: %s", parts[0])
+		}
+	}
+
+	// Parse the query attributes:
+	for _, attr := range qAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if len(parts) == 2 {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		} else {
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Query Attribute: %s", parts[0])
+		}
+		switch parts[0] {
+		case "pin-source":
+			pk11uri.PinSource = value
+		case "pin-value":
+			pk11uri.PinValue = value
+		case "module-name":
+			pk11uri.ModuleName = value
+		case "module-path":
+			pk11uri.ModulePath = value
+		case "max-sessions":
+			maxSessions, _ := strconv.ParseUint(value, 10, 32)
+			pk11uri.MaxSessions = int(maxSessions)
+		default:
+			return nil, fmt.Errorf("Unrecognized PKCS#11 URI Query Attribute: %s", parts[0])
+		}
+	}
+
+	return &pk11uri, nil
+}
+
+// LoadPKCS11Signer attempts to load a Signer given a PKCS11URI object that
+// identifies a key pair. At least three attributes must be specified:
+//
+//	Module:	use ModulePath to locate the PKCS #11 module library.
+//	Token:	use Serial or Token to specify the PKCS #11 token.
+//	Slot:	use SlotID, ID, or Object to specify the PKCS #11 key pair.
+//
+// For certain modules, a query attribute max-sessions is required in order to
+// prevent openning too many sessions to the module. Certain additional attributes,
+// such as pin-value, may be necessary depending on the situation. Refer to the
+// documentation for your PKCS #11 module for more details.
+//
+// An error is returned if the crypto11 module cannot find the module, token,
+// or the specified object.
```

Comment:
I still don't really know what this function is doing. Try connecting the HSM-related terminology to other things I might know about. Is a module path a unix socket? Is a token a password?

Reply:
You need to know what is PKCS #11. I don't think this package is the right place to document that. You can read the documentation on the `pkcs11` package for that.